abstract-factory/src/main/java/com/iluwatar/abstractfactory/App.java

```diff
@@ -41,105 +40,42 @@
  * and its implementations ( {@link ElfKingdomFactory}, {@link OrcKingdomFactory}). The example uses
  * both concrete implementations to create a king, a castle and an army.
  */
-public class App {
+@Slf4j
+public class App implements Runnable {
 
-  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);
-
-  private King king;
-  private Castle castle;
-  private Army army;
-
-  /**
-   * Creates kingdom.
-   */
-  public void createKingdom(final KingdomFactory factory) {
-    setKing(factory.createKing());
-    setCastle(factory.createCastle());
-    setArmy(factory.createArmy());
-  }
-
-  King getKing(final KingdomFactory factory) {
-    return factory.createKing();
-  }
-
-  public King getKing() {
-    return king;
-  }
-
-  private void setKing(final King king) {
-    this.king = king;
-  }
-
-  Castle getCastle(final KingdomFactory factory) {
-    return factory.createCastle();
-  }
-
-  public Castle getCastle() {
-    return castle;
-  }
-
-  private void setCastle(final Castle castle) {
-    this.castle = castle;
-  }
-
-  Army getArmy(final KingdomFactory factory) {
-    return factory.createArmy();
-  }
-
-  public Army getArmy() {
-    return army;
-  }
-
-  private void setArmy(final Army army) {
-    this.army = army;
-  }
-
-  /**
-   * The factory of kingdom factories.
-   */
-  public static class FactoryMaker {
-
-    /**
-     * Enumeration for the different types of Kingdoms.
-     */
-    public enum KingdomType {
-      ELF, ORC
-    }
-
-    /**
-     * The factory method to create KingdomFactory concrete objects.
-     */
-    public static KingdomFactory makeFactory(KingdomType type) {
-      switch (type) {
-        case ELF:
-          return new ElfKingdomFactory();
-        case ORC:
-          return new OrcKingdomFactory();
-        default:
-          throw new IllegalArgumentException("KingdomType not supported.");
-      }
-    }
-  }
+  @Getter
+  private final Kingdom kingdom = new Kingdom();
 
   /**
    * Program entry point.
    *
    * @param args command line args
    */
   public static void main(String[] args) {
-
     var app = new App();
+  }
 
-    LOGGER.info("Elf Kingdom");
-    app.createKingdom(FactoryMaker.makeFactory(KingdomType.ELF));
-    LOGGER.info(app.getArmy().getDescription());
-    LOGGER.info(app.getCastle().getDescription());
-    LOGGER.info(app.getKing().getDescription());
+  @Override
+  public void run() {
+    log.info("Elf Kingdom");
+    createKingdom(Kingdom.FactoryMaker.makeFactory(Kingdom.FactoryMaker.KingdomType.ELF));
```

Comment:
Maybe it would be even more clear if `createKingdom` would take only the race as a parameter and handle the factory making internally?

Reply:
I also think this option is more appropriate